pkg/liqo-controller-manager/shadowendpointslice-controller/shadowendpointslice_controller.go

```diff
@@ -0,0 +1,244 @@
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
+	vkv1alpha1 "github.com/liqotech/liqo/apis/virtualkubelet/v1alpha1"
+	liqoconsts "github.com/liqotech/liqo/pkg/consts"
+	clientutils "github.com/liqotech/liqo/pkg/utils/clients"
+	foreigncluster "github.com/liqotech/liqo/pkg/utils/foreignCluster"
+	"github.com/liqotech/liqo/pkg/virtualKubelet/forge"
+)
+
+const ctrlFieldManager = "shadow-endpointslice-controller"
+
+// Reconciler reconciles a ShadowEndpointSlice object.
+type Reconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=virtualkubelet.liqo.io,resources=shadowendpointslices,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=discovery.k8s.io,resources=endpointslices,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=discovery.liqo.io,resources=foreignclusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=discovery.liqo.io,resources=foreignclusters/status,verbs=get;list;watch
+
+// Reconcile ShadowEndpointSlices objects.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	nsName := req.NamespacedName
+	klog.V(4).Infof("reconcile shadowendpointslice %s", nsName)
+
+	var shadowEps vkv1alpha1.ShadowEndpointSlice
+	if err := r.Get(ctx, nsName, &shadowEps); err != nil {
+		if errors.IsNotFound(err) {
+			klog.V(4).Infof("shadowendpointslice %s not found", nsName)
+			return ctrl.Result{}, nil
+		}
+		klog.Errorf("an error occurred while getting shadowendpointslice %s: %v", nsName, err)
+		return ctrl.Result{}, err
+	}
+
+	// Get ForeignCluster associated with the shadowendpointslice
+	clusterID, ok := shadowEps.Labels[forge.LiqoOriginClusterIDKey]
+	if !ok {
+		klog.Errorf("shadowendpointslice %s has no label with origin clusterID", nsName)
+		return ctrl.Result{}, fmt.Errorf("shadowendpointslice %s has no label with origin clusterID", nsName)
+	}
+	fc, err := foreigncluster.GetForeignClusterByID(ctx, r.Client, clusterID)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Error("foreigncluster associated with shadowendpointslice %s not found", nsName)
+			return ctrl.Result{}, err
+		}
+		klog.Errorf("an error occurred while getting foreigncluster %s: %v", clusterID, err)
+		return ctrl.Result{}, err
+	}
+
+	// Check network status of the foreigncluster
+	networkStatus := foreigncluster.IsNetworkingEstablished(fc)
```

Comment:
Open question: what we should do when the networking is extrernal? Should we use the more generic function (not merged yet)? https://github.com/liqotech/liqo/blob/877d0b92babbcedc90f0265700d4f53408a5fe69/pkg/utils/foreignCluster/peeringStatus.go#L80